File: lib/ai/level-generator.ts
import { AIProviderFactory } from './providers';
import { AIMessage } from './types';
import { LessonStep } from '@/types';

export interface LevelContentInput {
  levelTitle: string;
  levelDescription: string;
  chapterTitle: string;
  material: string;
  difficulty: string;
  userFeedback?: {
    difficulty: 'too_hard' | 'just_right' | 'too_easy' | null;
    feedbackText: string;
  };
  previousAnswers?: {
    stepId: string;
    answer: string;
    isCorrect: boolean;
  }[];
}

function adjustDifficulty(baseDifficulty: string, feedback?: { difficulty: 'too_hard' | 'just_right' | 'too_easy' | null }): string {
  if (!feedback || !feedback.difficulty) return baseDifficulty === 'beginner' ? '初级' : baseDifficulty === 'advanced' ? '高级' : '中级';
  
  const diff = feedback.difficulty;
  if (diff === 'too_hard') {
    return baseDifficulty === 'advanced' ? '中级' : '初级';
  } else if (diff === 'too_easy') {
    return baseDifficulty === 'beginner' ? '中级' : '高级';
  }
  return baseDifficulty === 'beginner' ? '初级' : baseDifficulty === 'advanced' ? '高级' : '中级';
}

export async function generateLevelContent(input: LevelContentInput): Promise<LessonStep[]> {
  const provider = AIProviderFactory.getProvider();

  if (!provider.isAvailable()) {
    throw new Error(`AI provider ${provider.name} is not configured`);
  }

  const messages: AIMessage[] = [
    {
      role: 'system',
      content: '你根据材料生成学习关卡内容。输出必须是有效的JSON格式。支持HTML、LaTeX公式和SVG动画。',
    },
    {
      role: 'user',
      content: `
根据以下材料生成一个学习关卡的详细内容。

关卡标题: ${input.levelTitle}
关卡描述: ${input.levelDescription}
章节: ${input.chapterTitle}
${input.userFeedback ? `
用户反馈:
- 难度评价: ${input.userFeedback.difficulty === 'too_hard' ? '太难了' : input.userFeedback.difficulty === 'too_easy' ? '太简单了' : '刚刚好'}
- 详细反馈: ${input.userFeedback.feedbackText || '无'}
` : ''}
${input.previousAnswers && input.previousAnswers.length > 0 ? `
用户上一关答题情况:
${input.previousAnswers.map(a => `- ${a.stepId}: ${a.answer} (${a.isCorrect ? '正确' : '错误'})`).join('\n')}
` : ''}
难度: ${adjustDifficulty(input.difficulty, input.userFeedback)}

相关材料:
${input.material}

生成内容要求:
1. 2-4个学习步骤(steps)
2. 第一步通常是 info 类型介绍概念
3. 中间步骤可以是 info 或 multiple_choice
4. 最后一步建议是 multiple_choice 用于检验学习
5. content 字段支持HTML、LaTeX公式和SVG动画
6. 使用 $formula$ 表示行内公式，$$formula$$ 表示行间公式
7. 可以嵌入SVG动画来解释概念
8. 可以使用HTML表格来展示信息

示例：
{
  "steps": [
    {
      "id": "step-1",
      "type": "info",
      "title": "梯度下降原理",
      "content": "梯度下降优化公式：$$\\theta_{new} = \\theta_{old} - \\alpha \\nabla J(\\theta)$$\\n\\n其中 $\\alpha$ 是学习率，$\\nabla J(\\theta)$ 是目标函数的梯度。"
    },
    {
      "id": "step-2",
      "type": "info",
      "title": "神经网络结构",
      "content": "<table>\\n<tr><th>层</th><th>神经元数</th></tr>\\n<tr><td>输入层</td><td>784</td></tr>\\n<tr><td>隐藏层</td><td>128</td></tr>\\n<tr><td>输出层</td><td>10</td></tr>\\n</table>"
    },
    {
      "id": "step-3",
      "type": "multiple_choice",
      "title": "梯度下降检验",
      "content": "根据以下SVG动画回答问题：<svg width=\\"200\\" height=\\"100\\" data-animated>\\n<circle cx=\\"50\\" cy=\\"50\\" r=\\"30\\" fill=\\"blue\\">\\n<animate attributeName=\\"cx\\" from=\\"50\\" to=\\"150\\" dur=\\"2s\\" repeatCount=\\"indefinite\\"/>\\n</circle>\\n</svg>",
      "question": "图中圆圈在做什么运动？",
      "options": [
        {"id": "a", "text": "上下运动", "isCorrect": false},
        {"id": "b", "text": "左右运动", "isCorrect": true},
        {"id": "c", "text": "圆形旋转", "isCorrect": false},
        {"id": "d", "text": "静止不动", "isCorrect": false}
      ],
      "hint": "观察 cx 属性的变化"
    }
  ]
}

禁止:
- 不要使用比喻和类比
- 直接陈述知识点
- 使用材料中的实际内容

输出格式:
{
  "steps": [
    {
      "id": "step-1",
      "type": "info",
      "title": "步骤标题",
      "content": "步骤内容（支持HTML、LaTeX、SVG）"
    },
    {
      "id": "step-2",
      "type": "multiple_choice",
      "title": "问题标题",
      "content": "问题描述",
      "question": "问题",
      "options": [
        {"id": "a", "text": "选项A", "isCorrect": false},
        {"id": "b", "text": "选项B", "isCorrect": true}
      ],
      "hint": "提示"
    }
  ]
}
`,
    },
  ];

  const response = await provider.generateCompletion(messages, {
    temperature: 0.7,
    maxTokens: 4000,
    responseFormat: 'json',
  });

  const content = response.content;
  if (!content) {
    throw new Error('Failed to generate level content');
  }

  let jsonContent = content;

  const jsonMatch = content.match(/\`\`\`json\n?([\s\S]*?)\n?\`\`\`/);
  if (jsonMatch) {
    jsonContent = jsonMatch[1];
  } else {
    const jsonStart = content.indexOf('{');
    const jsonEnd = content.lastIndexOf('}');
    if (jsonStart !== -1 && jsonEnd !== -1 && jsonEnd > jsonStart) {
      jsonContent = content.substring(jsonStart, jsonEnd + 1);
    }
  }

  try {
    const result = JSON.parse(jsonContent);
    return result.steps;
  } catch (parseError: any) {
    let fixedContent = jsonContent.replace(/,(\s*[}\]])/g, '$1');
    try {
      const result = JSON.parse(fixedContent);
      return result.steps;
    } catch (fixError) {
      throw new Error(`Failed to parse AI response: ${parseError.message}`);
    }
  }
}

File: backend/services/level_generator.py
import json
import re
from typing import Dict, List, Any, Optional


class LevelGenerator:
    """关卡内容生成器"""
    
    def __init__(self, ai_provider):
        self.ai = ai_provider
    
    async def generate_content(
        self,
        level_title: str,
        level_description: str,
        chapter_title: str,
        material: str,
        knowledge_context: str = "",
        difficulty: str = "intermediate",
        user_context: str = "",
        previous_answers: List[Dict] = None,
        user_feedback: str = None
    ) -> List[Dict]:
        """生成关卡内容 - 支持上下文感知"""
        
        if previous_answers is None:
            previous_answers = []
        
        difficulty_label = {
            "beginner": "初级",
            "intermediate": "中级",
            "advanced": "高级"
        }.get(difficulty, "中级")
        
        system_prompt = """你是一个JSON生成器。只输出有效JSON，不要输出其他内容。
JSON中的文字必须是干净的UTF-8中文，不要有转义字符或HTML实体。"""
        
        # 构建上下文感知提示
        context_section = ""
        if user_context:
            context_section += f"\n用户学习上下文:\n{user_context}\n"
        
        # 分析用户答题表现
        if previous_answers and len(previous_answers) > 0:
            correct_count = sum(1 for a in previous_answers if a.get("isCorrect", False) or a.get("is_correct", False))
            total_count = len(previous_answers)
            accuracy = (correct_count / total_count * 100) if total_count > 0 else 0
            
            context_section += f"\n用户最近答题表现: {correct_count}/{total_count} 正确 ({accuracy:.0f}% 正确率)\n"
            
            if accuracy < 50:
                context_section += "注意: 用户正确率较低，需要更详细地解释概念，题目可以稍微简单一些\n"
            elif accuracy > 80:
                context_section += "注意: 用户正确率很高，可以增加一些有挑战性的题目\n"
        
        # 分析用户反馈
        if user_feedback:
            if "太难" in user_feedback or "too_hard" in user_feedback.lower():
                context_section += "用户反馈: 内容偏难，需要适当简化解释\n"
            elif "太简单" in user_feedback or "too_easy" in user_feedback.lower():
                context_section += "用户反馈: 内容偏简单，可以增加深度\n"
            else:
                context_section += f"用户反馈: {user_feedback}\n"
        
        user_prompt = f"""生成{difficulty_label}难度的学习关卡JSON。

关卡标题：{level_title}
关卡描述：{level_description}
章节标题：{chapter_title}

{context_section}
学习材料（重点根据此材料出题）：
{material[:2000]}

{f"参考知识库内容:\n{knowledge_context}" if knowledge_context else ""}

要求：
1. 关卡内容必须与关卡标题和材料相关
2. 每个关卡包含2-3个steps
3. steps必须包含info和quiz类型
4. quiz使用multiple_choice格式
5. 确保每个关卡内容独特，与其他关卡不重复
6. 根据用户的答题表现和反馈调整内容难度和深度

输出格式（必须是干净的有效JSON）：
{{"steps":[{{"id":"s1","type":"info","title":"标题","content":"内容"}},{{"id":"s2","type":"multiple_choice","title":"标题","question":"问题","options":[{{"id":"a","text":"选项A","isCorrect":true}},{{"id":"b","text":"选项B","isCorrect":false}}],"hint":"提示"}}]}}"""
        
        messages = [
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": user_prompt},
        ]
        
        response = self.ai.generate_completion(messages, {
            "temperature": 0.7,
            "max_tokens": 4000,
            "response_format": "json"
        })
        
        content = response.get("content", "")
        if not content:
            raise Exception("Failed to generate level content")
        
        # 提取JSON
        json_str = self._extract_json(content)
        
        try:
            result = json.loads(json_str)
            return result.get("steps", [])
        except json.JSONDecodeError as e:
            raise Exception(f"Failed to parse level content: {e}")
    
    def _extract_json(self, text: str) -> str:
        """从文本中提取JSON"""
        # 尝试从代码块中提取
        match = re.search(r'```(?:json)?\s*([\s\S]*?)\s*```', text)
        if match:
            json_str = match.group(1)
            # 清理Unicode转义
            json_str = self._clean_unicode(json_str)
            return json_str
        
        # 尝试直接提取
        start = text.find('{')
        end = text.rfind('}')
        if start != -1 and end != -1 and end > start:
            json_str = text[start:end+1]
            json_str = self._clean_unicode(json_str)
            return json_str
        
        return text
    
    def _clean_unicode(self, text: str) -> str:
        """清理Unicode转义"""
        # 解码 \uXXXX
        text = re.sub(r'\\u([0-9a-fA-F]{4})', lambda m: chr(int(m.group(1), 16)), text)
        # 清理转义字符
        text = text.replace('\\n', '\n').replace('\\r', '\r').replace('\\t', '\t')
        return text
